### PR TITLE
MM-47104 : Migrate "components/channel_notifications_modal/components/notification_section.jsx" to Typescript

### PR DESCRIPTION
**Pull request by aliziyacevik**
```release-note
SUMMARY
fixes https://github.com/mattermost/mattermost-server/issues/21065
```

### CODE REVIEW

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx [hunk 1]
```diff
@@ -1,90 +1,49 @@
...
-         * Update function to to expand or collapse a section
-         */
-        onUpdateSection: PropTypes.func.isRequired,
-
-        /**
-         * Error string from the server
-         */
-        serverError: PropTypes.string,
-    };
+type Props = {
```
Comment (sinansonmez) [inline webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx:11]: Can we keep the comments? If I remember correctly, I was told that these comments are useful for documentation

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx [hunk 2]
```diff
@@ -1,90 +1,49 @@
...
 
-    handleOnChangeThreads = (e) => {
+    handleOnChangeThreads = (e: ChangeEvent<HTMLInputElement>) => {
         const value = e.target.checked ? NotificationLevels.ALL : NotificationLevels.MENTION;
-
-        this.props.onChangeThreads(value);
-    };
+        if (this.props.onChangeThreads) {
+            this.props.onChangeThreads(value);
+        }
```
Comment (sinansonmez) [inline webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx:37]: You can avoid additional if statement by optional chaining. it is basically same is checking if it is null/ undefined with an if statement
```suggestion
        this.props.onChangeThreads?.(value);
```

**Review by sinansonmez — CHANGES_REQUESTED** [review]
Thanks for your effort @aliziyacevik Overall it looks good. I just have one minor comment to improve readability. Can you also migrate related test file as mentioned in the issue?

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx [hunk 3]
```diff
@@ -31,7 +31,7 @@ describe('components/channel_notifications_modal/NotificationSection', () => {
     });
 
     test('should match snapshot, DESKTOP on expanded view', () => {
-        const props = {...baseProps, expand: true};
+        const props= {...baseProps, expand: true};
```
Comment (sinansonmez) [inline webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx:34]: nit: I would keep the space before equal sign as most probably style check will complain about it.

**Review by sinansonmez — DISMISSED** [review]
LGTM 🚀

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx (hunk 3 above)
Comment (aliziyacevik) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx:34]: Thanks @sinansonmez, I added 2 more commits for typos and precaution for style check.

Comment (aliziyacevik) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx:34]: Didn't see your approvement before making changes, sorry  :(

Comment (sinansonmez) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx:34]: No worries at all. However, in this file, I was trying to say we should keep the space before equal like `const wrapper = shallow`. I have seen that you have removed all spaces before equal sign. Can you revert that? Sorry for the confusion

Comment (aliziyacevik) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.test.tsx:34]: Haha, my bad.  on it.

**Review by sinansonmez — DISMISSED** [review]

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx (hunk 1 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx:11]: Agree with Sinan’s point

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Pending comments change; otherwise looks good to me.

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx (hunk 1 above)
Comment (aliziyacevik) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx:11]: I added them back.

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx [hunk 4]
```diff
@@ -1,90 +1,91 @@
...
-
-        this.props.onChangeThreads(value);
-    };
+        this.props.onChangeThreads?.(value);
+    }
 
     handleExpandSection = () => {
         this.props.onUpdateSection(this.props.section);
-    };
+    }
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx:84]: Eslint will complain of semicolon here and can you please check other CI failures

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Comment by M-ZubairAhmed** [comment]
Removing QA since its just type changes

File: webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx (hunk 4 above)
Comment (aliziyacevik) [reply webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx]: done. @M-ZubairAhmed.

**Review by M-ZubairAhmed — APPROVED** [review]

**Review by hmhealey — CHANGES_REQUESTED** [review]
We're having some type check failures with these (see https://github.com/mattermost/mattermost-server/actions/runs/4667189100/jobs/8262675763?pr=22897). Could you check those out?

The ones in `notification_section.test.tsx` are likely because TS can't figure out the type of `wrapper` correctly. If you change those declarations to `const wrapper = shallow<NotificationSection>(...):`, I think it should fix those.

For the ones in `channel_notifications_modal.tsx`, the errors are a bit cryptic, but I think it seems like there's function being passed as a prop that has an incorrect type

**Comment by hmhealey** [comment]
Hi @aliziyacevik. Do you have any questions about my feedback or need any help with it?

**Comment by aliziyacevik** [comment]
@hmhealey sorry for late notice. No, thanks I dont have any questions. I believe I can't do further updates on this.  If someone takes over this issue please feel free I dont want to occupy this issue since I'm assigned to this.

**Pull request closed without merging**